core/src/http_helpers.rs

```diff
@@ -47,33 +54,45 @@ pub async fn read_body(
 		return Err(GenericTransportError::TooLarge);
 	}
 
-	let first_chunk =
-		body.next().await.ok_or(GenericTransportError::Malformed)?.map_err(GenericTransportError::Inner)?;
+	futures_util::pin_mut!(body);
 
-	if first_chunk.len() > max_request_body_size as usize {
-		return Err(GenericTransportError::TooLarge);
-	}
+	let mut received_data = Vec::with_capacity(body_size as usize);
+	let mut is_single = None;
 
-	let first_non_whitespace = first_chunk.iter().find(|byte| !byte.is_ascii_whitespace());
+	while let Some(d) = body.data().await {
+		let data = d.map_err(|e| GenericTransportError::Inner(anyhow!(e.into())))?;
 
-	let single = match first_non_whitespace {
-		Some(b'{') => true,
-		Some(b'[') => false,
-		_ => return Err(GenericTransportError::Malformed),
-	};
+		if data.chunk().len() > max_request_body_size as usize {
+			return Err(GenericTransportError::TooLarge);
+		}
 
-	let mut received_data = Vec::with_capacity(body_size as usize);
-	received_data.extend_from_slice(&first_chunk);
+		if received_data.is_empty() {
+			let first_non_whitespace = data.chunk().iter().find(|byte| !byte.is_ascii_whitespace());
```

Comment:
Nit: If we were to remove the previous check we could optimize for malformed input with an `.take(32)`, if we don't see the `{` / `[` character we could bail out sooner

Reply:
let's do it another PR, we could also fetch the position here skip the bytes with whitespaces

Reply:
I changed my mind, I used `take(128)` but not sure what is reasonable limit

Reply:
That sounds reasonable to me. I'm not sure if the spec says anything about it, but I would expect most cases to be a few characters away

Reply:
it's just that it's opinionated and we might reject valid HTTP responses that contains more than 128 whitespaces but it's valid JSON-RPC response :P 

however, I think it's fair trade-off because the JSON decoding happens later.